### PR TITLE
feat(api): megaparse under api

### PR DESCRIPTION
**Pull request by StanGirard**
This commit updates the dependencies in the `pyproject.toml` file. It adds new packages such as `fastapi`, `uvicorn`, `ratelimit`, `requests`, `backoff`, `pypdf`, `psutil`, and `numpy`. Additionally, it updates the version of `pycryptodome` to 3.21.0.

### CODE REVIEW

**Comment by chloedia** [comment]
Left : 
[ ] Tests on new megaparse
[ ] Fix Read me and evaluation notebook

File: megaparse/checker/format_checker.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    improve_layout(elements: List[Element]) -> List[Element]
+        Processes a list of elements, converting HTML tables to markdown tables and improving the overall layout.
+
+    """
+
+    def __init__(self, model: BaseChatModel):
+        self.model = model
+
+    def check(self, elements: List[Element]):
+        return "ok"
```
Comment (StanGirard) [inline megaparse/checker/format_checker.py:25]: Pk ?

File: megaparse/checker/markdown_processor.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+#             # other_pages_paragraphs = self.split_into_paragraphs(pages[1:])
+
+#             cleaned_paragraphs, duplicate_paragraphs = self.remove_duplicates(
+#                 paragraphs
+#             )
+#             header_components_count = self.identify_header_components(
+#                 duplicate_paragraphs
+#             )
+
+#             if self.strict:
```
Comment (StanGirard) [inline megaparse/checker/markdown_processor.py:198]: Remove the deadcode

File: megaparse/main.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+from pathlib import Path
+from langchain_openai import ChatOpenAI
+
+from megaparse.parser.unstructured_parser import UnstructuredParser
+
+from megaparse.parser import MegaParser
+from megaparse.checker.format_checker import FormatChecker
+
+# Dead Code, now relying on unstructured for all the other formats
+# class Converter:
```
Comment (StanGirard) [inline megaparse/main.py:12]: remove dead code

File: tests/docx/test_docx_processing.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+@pytest.mark.asyncio
+async def test_megaparse_docx_processor_fail():
+    p = Path("./tests/docx/sample.docx")
+    parser = LlamaParser(api_key=str(uuid4()))
+    processor = MegaParse(parser=parser)
+    with pytest.raises(ValueError):
+        await processor.aload(file_path=p)
+
+    parser = MegaParseVision(model=ChatOpenAI(model="gpt-4o", api_key=str(uuid4())))  # type: ignore
```
Comment (StanGirard) [inline tests/docx/test_docx_processing.py:28]: ouch ca va nous couter chère. Mais en place un monkey patch pour fake ChatOpenAI. Langchain le fait

File: tests/epub/test_epub_processing.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+@pytest.mark.asyncio
+async def test_megaparse_epub_processor_fail():
+    p = Path("./tests/epub/Sway.epub")
+    parser = LlamaParser(api_key=str(uuid4()))
+    processor = MegaParse(parser=parser)
+    with pytest.raises(ValueError):
+        await processor.aload(file_path=p)
+
+    parser = MegaParseVision(model=ChatOpenAI(model="gpt-4o", api_key=str(uuid4())))  # type: ignore
```
Comment (StanGirard) [inline tests/epub/test_epub_processing.py:28]: idem là. Il faut monkey patch pour éviter de call pour de vrai

File: tests/html/test_html_processing.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+@pytest.mark.asyncio
+async def test_megaparse_html_processor_fail():
+    p = Path("./tests/html/sample_complexe.html")
+    parser = LlamaParser(api_key=str(uuid4()))
+    processor = MegaParse(parser=parser)
+    with pytest.raises(ValueError):
+        await processor.aload(file_path=p)
+
+    parser = MegaParseVision(model=ChatOpenAI(model="gpt-4o", api_key=str(uuid4())))  # type: ignore
```
Comment (StanGirard) [inline tests/html/test_html_processing.py:28]: idem. Monkey

File: tests/odt/test_odt_processing.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+@pytest.mark.asyncio
+async def test_megaparse_odt_processor_fail():
+    p = Path("./tests/odt/file-sample_500kB.odt")
+    parser = LlamaParser(api_key=str(uuid4()))
+    processor = MegaParse(parser=parser)
+    with pytest.raises(ValueError):
+        await processor.aload(file_path=p)
+
+    parser = MegaParseVision(model=ChatOpenAI(model="gpt-4o", api_key=str(uuid4())))  # type: ignore
```
Comment (StanGirard) [inline tests/odt/test_odt_processing.py:28]: Idem, monkey

File: tests/pptx/test_pptx_processing.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+@pytest.mark.asyncio
+async def test_megaparse_pptx_processor_fail():
+    p = Path("./tests/pptx/sample.pptx")
+    parser = LlamaParser(api_key=str(uuid4()))
+    processor = MegaParse(parser=parser)
+    with pytest.raises(ValueError):
+        await processor.aload(file_path=p)
+
+    parser = MegaParseVision(model=ChatOpenAI(model="gpt-4o", api_key=str(uuid4())))  # type: ignore
```
Comment (StanGirard) [inline tests/pptx/test_pptx_processing.py:28]: idem, monkey

File: tests/xls/test_xls_processing.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+@pytest.mark.asyncio
+async def test_megaparse_xls_processor_fail():
+    p = Path("./tests/xls/file_example_XLS_50.xls")
+    parser = LlamaParser(api_key=str(uuid4()))
+    processor = MegaParse(parser=parser)
+    with pytest.raises(ValueError):
+        await processor.aload(file_path=p)
+
+    parser = MegaParseVision(model=ChatOpenAI(model="gpt-4o", api_key=str(uuid4())))  # type: ignore
```
Comment (StanGirard) [inline tests/xls/test_xls_processing.py:36]: idem, monkey

**Comment by StanGirard** [comment]
Also update documentation to explain how to work with the API.

File: megaparse/checker/format_checker.py (hunk 1 above)
Comment (chloedia) [reply megaparse/checker/format_checker.py:25]: Je vais implémenter plus tard !

File: megaparse/checker/markdown_processor.py (hunk 2 above)
Comment (chloedia) [reply megaparse/checker/markdown_processor.py:198]: It's not dead code it is something to keep for when i tackle FormatCheckers

File: megaparse/parser/unstructured_parser.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        table_stack = []
+
+        if "emphasized_text_contents" in metadata:
+            print(metadata["emphasized_text_contents"])
+
+        # Markdown line defaults to empty
+        markdown_line = ""
+
+        # Element type-specific markdown content
+        markdown_types = {
```
Comment (AmineDiro) [inline megaparse/parser/unstructured_parser.py:45]: Define enum of markdown types

File: tests/xls/test_xls_processing.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    result = await processor.aload(file_path=p)
+    assert len(result) > 0
+
+
+@pytest.mark.asyncio
+async def test_megaparse_xlsx_processor():
+    p = Path("./tests/xls/file_example_XLSX_50.xlsx")
+    processor = MegaParse()
+    result = await processor.aload(file_path=p)
+    assert len(result) > 0
```
Comment (AmineDiro) [inline tests/xls/test_xls_processing.py:26]: We should have a longer list of assertion here to verify parsing has worked the way we want

File: tests/docx/test_docx_processing.py
Comment (AmineDiro) [inline tests/docx/test_docx_processing.py:1]: Example of listing all files in directory and parametrizing a fixture with the documents1
```python
def list_files_in_data_directory():
    data_dir = Path("./tests/data")
    return [file for file in data_dir.iterdir() if file.is_file()]


def file_id(file_path):
    return file_path.name


@pytest.fixture(params=list_files_in_data_directory(), ids=file_id)
def qfile_placeholder(request) -> QuivrFile:
    """Fixture to create a QuivrFile for each file in data/ directory."""
    file_path = request.param
    extension = FileExtension(file_path.suffix)

    return QuivrFile(
        id=uuid4(),
        file_extension=extension,
        original_filename=file_path.name,
        file_sha1="124",  # Replace with a real SHA1 calculation if needed
        file_size=file_path.stat().st_size,
        path=file_path,
    )
```

```jsonl
{
"file_path" : "./data/test.pdf",
"extension": ".pdf",
"expected" :{
"length": ...,
...
}
{
"file_path" : "./data/test.pdf",
"extension": ".pdf",
"expected" :{
"length": ...,
...
}


}
```

**Review by AmineDiro — CHANGES_REQUESTED** [review]

File: megaparse/api/general.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+from langchain_openai import ChatOpenAI
+from langchain_anthropic import ChatAnthropic
+from llama_parse.utils import Language
+
+
+app = FastAPI()
+
+
+@app.get("/")
+def read_root():
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/api/general.py:20]: Like this ?

File: megaparse/api/general.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    check_table=False,
+    language: Language = Language.ENGLISH,
+    parsing_instruction: str | None = None,
+    model_name: str | None = None,
+):
+    _check_free_memory()
+
+    model = None
+    if model_name:
+        if "gpt" in model_name:
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/api/general.py:50]: what do you mean ?

File: megaparse/api/general.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+                api_key=os.getenv("ANTHROPIC_API_KEY"),  # type: ignore
+                timeout=60,
+                stop=None,
+            )
+
+        else:
+            raise HTTPException(
+                status_code=400, detail="Model not supported for MegaParse Vision"
+            )
+
```
Comment (chloedia) [inline megaparse/api/general.py:64]: Do I raise it here? Or no raises ?

File: megaparse/parser/unstructured_parser.py (hunk 10 above)
Comment (chloedia) [reply megaparse/parser/unstructured_parser.py:45]: Not sure to understand, why isn't the dict better ?

Comment (chloedia) [reply megaparse/parser/unstructured_parser.py:45]: You mean smth like that ? : 
class MarkDownType(str, Enum):
    """Markdown type enumeration."""

    TITLE = "Title"
    SUBTITLE = "Subtitle"
    HEADER = "Header"
    FOOTER = "Footer"
    NARRATIVE_TEXT = "NarrativeText"
    LIST_ITEM = "ListItem"
    TABLE = "Table"
    PAGE_BREAK = "PageBreak"
    IMAGE = "Image"
    FORMULA = "Formula"
    FIGURE_CAPTION = "FigureCaption"
    ADDRESS = "Address"
    EMAIL_ADDRESS = "EmailAddress"
    CODE_SNIPPET = "CodeSnippet"
    PAGE_NUMBER = "PageNumber"
    DEFAULT = "Default"
    UNDEFINED = "Undefined"

and then a dict ? Just to make sure there are no new types ?

File: tests/docx/test_docx_processing.py
Comment (chloedia) [reply tests/docx/test_docx_processing.py:1]: I created a ticket (#Core-266)

File: tests/xls/test_xls_processing.py (hunk 11 above)
Comment (chloedia) [reply tests/xls/test_xls_processing.py:26]: Same created a test ticket #core-266

File: megaparse/parser/type.py
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/parser/type.py:1]: Do you mean renaming the class or the file ? If the file should I split in two files ?

File: megaparse/api/general.py
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/api/general.py:1]: I'll rename it after the solving ot the other comments sinc it deletes the comments (or I don't get where they go)

File: megaparse/api/general.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+            api_key=str(os.getenv("LLAMA_CLOUD_API_KEY")),
+            language=language,  # type: ignore
+            parsing_instruction=parsing_instruction,
+        ),
+        "megaparse_vision": MegaParseVision(model=model),
+    }
+
+    with open(file.filename, "wb") as f:  # type: ignore
+        f.write(file.file.read())
+        megaparse = MegaParse(parser=parser_dict[method])
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/api/general.py:79]: I don't know, I like the fact that you can set up first your parser and then you can just use it.

File: megaparse/api/general.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        result = await megaparse.aload(file_path=str(file.filename))
+        os.remove(file.filename)  # type: ignore
+        return {"message": "File uploaded successfully", "result": result}
+
+
+@app.post("/url")
+async def upload_url(url: str):
+    loader = PlaywrightURLLoader(urls=[url], remove_selectors=["header", "footer"])
+
+    if url.endswith(".pdf"):
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/api/general.py:89]: better ?

File: megaparse/checker/format_checker.py [hunk 17]
```diff
@@ -0,0 +1,25 @@
+from typing import List
+from langchain_core.language_models.chat_models import BaseChatModel
+from unstructured.documents.elements import Element
+
+
+# TODO: Implement the FormatChecker class @Chloe
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/checker/format_checker.py:6]: added in type.py

File: megaparse/main.py [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        self.last_parsed_document: str = ""
+
+    async def aload(self, file_path: Path | str) -> str:
+        if isinstance(file_path, str):
+            file_path = Path(file_path)
+        file_extension: str = file_path.suffix
+
+        if file_extension != ".pdf":
+            if self.format_checker:
+                raise ValueError(
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/main.py:29]: Format Checker is something help, it is more like a module to post process the parsing

File: megaparse/main.py [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+        self.parser = parser
+        self.format_checker = format_checker
+        self.last_parsed_document: str = ""
+
+    async def aload(self, file_path: Path | str) -> str:
+        if isinstance(file_path, str):
+            file_path = Path(file_path)
+        file_extension: str = file_path.suffix
+
+        if file_extension != ".pdf":
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/main.py:27]: Done in type.py

File: megaparse/main.py
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/main.py:1]: Will be done after resolving the other issues

File: megaparse/main.py [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+        self.last_parsed_document = parsed_document
+        return parsed_document
+
+    def save(self, file_path: Path | str) -> None:
+        os.makedirs(os.path.dirname(file_path), exist_ok=True)
+        with open(file_path, "w+") as f:
+            f.write(self.last_parsed_document)
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/main.py:91]: Is it bad pratice to live a script to see how it is used ?

File: megaparse/api/general.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/api/general.py:91]: We need to know if we dl the pdf or not

File: megaparse/main.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/main.py:91]: until ot in standard agree to disagree (Stan)

File: megaparse/main.py
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/main.py:1]: rename file core/__init__.py

File: megaparse/parser/megaparse_vision.py [hunk 21]
```diff
@@ -64,14 +61,22 @@ class TagEnum(str, Enum):
...
-            raise NotImplementedError("Claude support not yet implemented")
-        else:
-            raise ValueError(f"Model {model} not supported")
+class MegaParseVision(MegaParser):
+    def __init__(self, model: BaseChatModel):
+        if hasattr(model, "model_name"):
+            assert (
+                model.model_name  # type: ignore
+                in [
+                    "gpt-4o",
```
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/parser/megaparse_vision.py:70]: no other choice (maybe lite llm later)

File: megaparse/parser/unstructured_parser.py (hunk 10 above)
Comment (chloedia) [reply megaparse/parser/unstructured_parser.py:45]: to see later, adds complexity (create a ticket)

File: megaparse/api/general.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (chloedia) [reply megaparse/api/general.py]: -> debate later

**Pull request merged**